src/miral/basic_window_manager.cpp

```diff
@@ -535,15 +535,27 @@ void miral::BasicWindowManager::focus_next_application()
             }
             while (focus_controller->focused_session() != prev.application());
         }
+        else
+        {
+            do
+            {
+                focus_controller->focus_next_session();
 
+                if (can_activate_window_for_session(focus_controller->focused_session()))
```

Comment:
Since the loop code seems to be the same in both this and the block above, it looks like you could move the `workspaces_containing_window.empty()` check to inside the loop. (not sure if you need to cache it to a local bool or not).

Reply:
The code is not the same.

```
                if (can_activate_window_for_session_in_workspace(
                    focus_controller->focused_session(),
                    workspaces_containing_window))
```
vs
```
                if (can_activate_window_for_session(focus_controller->focused_session()))
```
Note the `workspaces_containing_window` parameter in the fist block.